src/type/result.ts

```diff
@@ -0,0 +1,74 @@
+export type Result<T, E extends string = string> =
+  | SuccessResult<T>
+  | FailureResult<E>;
+export const success = <T>(value: T): SuccessResult<T> => ({ ok: true, value });
+export type SuccessResult<T> = { ok: true; value: T };
+type Failure = {
+  (error: Error): FailureResult<typeof unknownCode>;
+  <C extends string | undefined>(code: C, error: Error): FailureResult<
+    C extends string ? C : typeof unknownCode
+  >;
+};
+const unknownCode = "UNKNOWN" as const;
+export const failure: Failure = <C extends string>(
+  codeOrError: C | undefined | Error,
+  error?: Error
+) => {
+  switch (typeof codeOrError) {
+    case "undefined": {
+      if (typeof error === "undefined") {
```

Comment:
下は`if (error === undefined) {`なので合わせたほうがきれいそう？
```suggestion
      if (error === undefined) {
```